File: FinalProject/Java/src/functionalProgram/Processor.java
package functionalProgram;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.Annotation;
import java.util.function.*;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;


public class Processor {

//TODO LINK TO STREAMS AND FUNCTIONAL https://www.geeksforgeeks.org/stream-in-java/
    public void process(String[] args) throws IOException {
        //set up grep ad wc
        boolean grep = false;
        boolean wc = false;
        String path = "";
        if(args[0].equals("grep")) {
            path = args[2];
            grep = true;
            if (args.length == 5)
                wc = true;
        }else{
            wc = true;
            path = args[1];
        }


        //get text
        File file = new File(path);
        List<String> lines = Files.readAllLines(Paths.get(path));

        //perform task based on grep/wc structure
        if(grep && !wc){ //basic grep ex. "grep this silly.txt"
            lines.stream().filter(line -> line.contains(args[1])).forEach(y->System.out.println(y));
        }
        else if(grep && wc){//complicated grep ex.  "grep this silly.txt | wc"
            Map myMap =
                lines.stream()
                    .filter(line -> line.contains(args[1]))
                    .map(line -> line.toLowerCase())
                    .flatMap(line -> Stream.of(line.split("\\s+")))//split
                    .map(word -> word.replaceAll("[^a-zA-Z]", "").trim())
                    .filter(word -> !word.equals(""))//get rid of "" that were counted as words
                    .map(word -> new AbstractMap.SimpleEntry<String,Integer>(word, 1))
                    .collect(Collectors.toMap(word -> word.getKey(), word2 -> word2.getValue(), (wordA, wordB) -> wordA+wordB));
            printMe(myMap);

        } else if(wc){//just wc. ex. "wc silly.text")
            Map WCMap =
                lines.stream()
                    .map(line -> line.toLowerCase())
                    .flatMap(line -> Stream.of(line.split("\\s+")))//split
                    .map(word -> word.replaceAll("[^a-zA-Z]", "").trim())
                    .filter(line -> !line.equals(""))//get rid of "" that were counted as words
                    .map(word -> new AbstractMap.SimpleEntry<String,Integer>(word, 1))
                    .collect(Collectors.toMap(word -> word.getKey(), word2 -> word2.getValue(), (wordA, wordB) -> wordA+wordB));
                    //.forEach(y->System.out.println(y));
            printMe(WCMap);

        }
    }
    public static void printMe(Map<String, Integer> map) {
        for(Map.Entry entry : map.entrySet()){
            System.out.println(entry.getKey()+" "+ entry.getValue());
        }
    }



    public static void main(String[] args) throws IOException {
        Processor processor = new Processor();
        processor.process(args);
    }
}

File: FinalProject/JavaScript/Functional/Processor.js
function printMe(ans) {
    for(var i in ans) {
        console.log(ans[i]);
    }
}
function printWC(ans) {
    for(i in ans){
        console.log(`${i}, ${ans[i]}`);
    }
}
class Processor{
    constructor(){
    }

    process(args){
        // console.log(args)
        //set up grep ad wc
        let grep = false;
        let wc = false;
        let path = ""
        let searchString = ""
        if(args[2] == "grep") {
            searchString = args[3];
            path = args[4];
            grep = true;
            if (args.length = 7 && args[6] == "wc")
                wc = true;
        }else{
            path = args[3];
            wc = true;
        }
        //set up text
        const fs = require('fs');
        path = "file:"+path;
        //console.log(path);
        const fileUrl = new URL(path);
        let text = fs.readFileSync(fileUrl).toString();
        // console.log(text);

        //
        let lines = text.split(/\n+/);
        //printMe(lines);
        //perform task based on grep/wc structure
        if(grep && !wc){ //basic grep ex. "grep this silly.txt"
            let ans = lines.filter(line => line.includes(searchString));
            printMe(ans);
        }
        else if(grep && wc){//complicated grep ex.  "grep this silly.txt | wc"
            let ans =
                lines.filter(line => line.includes(searchString))
                    .reduce((text, word) => text+word, " ").toLowerCase().split(/\s/g)
                    .map(word => word.replace(/[^a-zA-Z_]/g, ''))
                    .filter(word => !(word == ""))//get rid of "" that were counted as words
                    .reduce(function(stats, word) {
                        if (stats[word]) {
                            stats[word] = stats[word] + 1;
                        } else {
                            stats[word] = 1;
                        }
                        return stats;
                    }, []);
            printWC(ans);
        } else if(wc) {//just wc. ex. "wc silly.text"
            let ans =
                lines
                    .reduce((text, word) => text + word, "").toLowerCase().split(/\s/g)
                    .map(word => word.replace(/[^a-zA-Z_]/g, ''))
                    .filter(word => !(word == ""))//get rid of "" that were counted as words
                    .reduce(function (stats, word) {
                        if (stats[word]) {
                            stats[word] = stats[word] + 1;
                        } else {
                            stats[word] = 1;
                        }
                        return stats;
                    }, []);
            printWC(ans);
        }
    }

}
let processor = new Processor();
processor.process(process.argv);
